library/L1_Peripheral/msp432p401r/system_controller.hpp

```diff
@@ -145,111 +226,158 @@ class SystemController final : public sjsu::SystemController
+      default: break;
+    }
 
-  /// Configures the digitally controlled clock (DCO) to produce the desired
-  /// target nominal frequency to drive the master clock.
-  ///
-  /// @param frequency The desired target frequency between 1 MHz and 48 MHz.
-  void SetSystemClockFrequency(
-      units::frequency::megahertz_t frequency) const
-  {
-    const DcoConfiguration_t kDcoConfiguration =
-        CalculateDcoConfiguration(frequency);
+    switch (clock_configuration_.master.clock_source)
+    {
+      case Oscillator::kLowFrequency: mclk = lfxt; break;
+      case Oscillator::kVeryLowFrequency: mclk = vlo; break;
+      case Oscillator::kReference: mclk = refo; break;
+      case Oscillator::kDigitallyControlled: mclk = dco; break;
+      case Oscillator::kModule: mclk = mod; break;
+      case Oscillator::kHighFrequency: mclk = hfxt; break;
+    }
 
-    UnlockClockSystemRegisters();
+    switch (clock_configuration_.subsystem_master.clock_source)
     {
-      clock_system->CTL0 = bit::Insert(clock_system->CTL0,
-                                       kDcoConfiguration.tuning_value,
-                                       Control0Register::kTuningSelect);
-      clock_system->CTL0 = bit::Insert(clock_system->CTL0,
-                                       kDcoConfiguration.frequency_select,
-                                       Control0Register::kFrequencySelect);
-      clock_system->CTL0 =
-          bit::Set(clock_system->CTL0, Control0Register::kEnable);
+      case Oscillator::kLowFrequency: smclk = lfxt; break;
+      case Oscillator::kVeryLowFrequency: smclk = vlo; break;
+      case Oscillator::kReference: smclk = refo; break;
+      case Oscillator::kDigitallyControlled: smclk = dco; break;
+      case Oscillator::kModule: smclk = mod; break;
+      case Oscillator::kHighFrequency: smclk = hfxt; break;
     }
-    LockClockSystemRegisters();
 
-    WaitForClockReadyStatus(Clock::kMaster);
-    SetClockSource(Clock::kMaster, Oscillator::kDigitallyControlled);
-    SetClockSource(Clock::kSubsystemMaster, Oscillator::kDigitallyControlled);
+    switch (clock_configuration_.backup.clock_source)
+    {
+      case Oscillator::kLowFrequency:
+      case Oscillator::kReference:
+      default: break;
+    }
 
-    speed_in_hertz = frequency;
-  }
+    constexpr std::array<uint32_t, 8> kDividers = {
+      1, 2, 4, 8, 16, 32, 64, 128
+    };
 
-  /// Sets the divider for the subsystem master clock (HSMCLK).
-  ///
-  /// @param peripheral_divider Divider value to set. Only the following
-  ///                           dividers are available: 1, 2, 4, 8, 16, 32, 64,
-  ///                           128.
-  void SetPeripheralClockDivider(PeripheralID,
-                                 uint8_t peripheral_divider) const
-  {
-    SetClockDivider(Clock::kSubsystemMaster, peripheral_divider);
+    auxiliary_clock_rate_ =
+        aclk / kDividers[Value(clock_configuration_.auxiliary.divider)];
```

Comment:
I think this would more optimal rather than using a look up table. It may be interesting to test this.
 
```suggestion
    auxiliary_clock_rate_ =
        aclk / (1 << Value(clock_configuration_.auxiliary.divider));
```

Lookup table option:
- Permant 4 x 8 bytes (32 bytes) 
- Array lookup

Shift left:
- Potentially a single 16-bit thumb instruction to perform the `1 << amount` arithmetic.